### PR TITLE
introduced 'Pending' state for backend's health

### PR DESCRIPTION
**Pull request by andythsu**
This commit adds a 'Pending' state to keep track of backend's health.

Resolves: https://github.com/trinodb/trino-gateway/issues/80

### CODE REVIEW

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/BackendHealthState.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.gateway.ha.clustermonitor;
+
+public enum BackendHealthState {
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/BackendHealthState.java:16]: Rename to BackendStatus.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java [hunk 2]
```diff
@@ -24,7 +24,7 @@ public class ClusterStats
     private int queuedQueryCount;
     private int blockedQueryCount;
     private int numWorkerNodes;
-    private boolean healthy;
+    private BackendHealthState healthy = BackendHealthState.UNHEALTHY;
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java:27]: Rename the variable to backendStatus.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java [hunk 3]
```diff
@@ -75,14 +75,29 @@ public void setNumWorkerNodes(int numWorkerNodes)
...
+        return this.healthy == BackendHealthState.HEALTHY;
+    }
+
+    public boolean isPending()
+    {
+        return this.healthy == BackendHealthState.PENDING;
     }
 
-    public void setHealthy(boolean healthy)
+    public boolean isUnhealthy()
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java:86]: Remove above isHealthy, isPending & isUnhealthy method.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java [hunk 4]
```diff
@@ -75,14 +75,29 @@ public void setNumWorkerNodes(int numWorkerNodes)
...
+    {
+        return this.healthy == BackendHealthState.UNHEALTHY;
+    }
+
+    public void setHealthy(BackendHealthState healthy)
     {
         this.healthy = healthy;
     }
 
+    public BackendHealthState getHealthy()
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStats.java:96]: Move to before the setter.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsHttpMonitor.java [hunk 5]
```diff
@@ -68,11 +68,11 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
 
-            clusterStats.setNumWorkerNodes((int) result.get("activeWorkers"));
-            clusterStats.setQueuedQueryCount((int) result.get("queuedQueries"));
-            clusterStats.setRunningQueryCount((int) result.get("runningQueries"));
-            clusterStats.setBlockedQueryCount((int) result.get("blockedQueries"));
-            clusterStats.setHealthy(clusterStats.getNumWorkerNodes() > 0);
+            clusterStats.setNumWorkerNodes((int) result.getOrDefault("activeWorkers", 0));
+            clusterStats.setQueuedQueryCount((int) result.getOrDefault("queuedQueries", 0));
+            clusterStats.setRunningQueryCount((int) result.getOrDefault("runningQueries", 0));
+            clusterStats.setBlockedQueryCount((int) result.getOrDefault("blockedQueries", 0));
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsHttpMonitor.java:74]: Why these changes are required for introducing pending status?

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java [hunk 6]
```diff
@@ -98,16 +98,16 @@ public ClusterStats monitor(ProxyBackendConfiguration backend)
...
             clusterStats.setRunningQueryCount(partialState.getOrDefault("RUNNING", 0));
             return clusterStats;
         }
         catch (TimeoutException e) {
-            log.error("timed out fetching status for {} backend", url, e);
+            log.error("timed out fetching status for {} backend", jdbcUrl, e);
         }
         catch (Exception e) {
-            log.error("could not fetch status for {} backend", url, e);
+            log.error("could not fetch status for {} backend", jdbcUrl, e);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsJdbcMonitor.java:110]: Extract a commit. These changes are unrelated to pending status.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java [hunk 7]
```diff
@@ -29,8 +29,7 @@ public HealthCheckObserver(RoutingManager routingManager)
     public void observe(java.util.List<ClusterStats> clustersStats)
     {
         for (ClusterStats clusterStats : clustersStats) {
-            routingManager.upateBackEndHealth(clusterStats.getClusterId(), clusterStats.isHealthy());
-            routingManager.updateBackEndHealthDB(clusterStats);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: Why did you remove this line?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/RoutingManager.java [hunk 8]
```diff
@@ -66,7 +66,7 @@ public String load(String queryId)
                                     }
                                 });
 
-        this.backendToHealth = new ConcurrentHashMap<String, Boolean>();
+        this.backendToHealth = new ConcurrentHashMap();
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/RoutingManager.java:69]: Avoid raw usage of parameterized map.

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/RoutingManager.java:69]: is it necessary to put `<String, BackendHealthState>` or just `<>()` is enough?

Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/RoutingManager.java:69]: `<>()` is enough.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java (hunk 7 above)
Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: this is the line that caused the problem. It directly manipulates the DB, which is unwanted. This method has been deleted in `routingManager`

**Review by willmostly — CHANGES_REQUESTED** [review]
This is moving in the right direction, but as stated in the comments I think this change needs to include an update to the DB model as well.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/BackendStatus.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.gateway.ha.clustermonitor;
+
+public enum BackendStatus {
+  HEALTHY,
+  PENDING,
```
[the remark this answers is not in the document]
Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/BackendStatus.java:18]: sure, will add a javadoc

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java (hunk 7 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: > having multiple gateways could lead to inconsistent states for sure

In deploying using Helm Chart and for HA purpose for gateway server itself, people would use multiple gateways. Our team too are using HPA for the gateway itself, with 3 in minimum. 

If the status is local with serveral servers, then everytime I refresh the page I could get different Healthy status right?

I think a consistent sync is necessary, like 
- if status is inactive then the monitor could query and check if it is healthy
  - if the result is unhealthy, show unhealthy (meaning that it will do health check later) 
  - if the result is pending, show pending (similar to unhealthy but pending could also mean it is in checking state?)
  - if the result is healthy, then update DB so no other gateway server linked to same db would do unnecessary querying
- if status is active, then monitor does nothing

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: by status do you mean backend status? I think you have active and inactive cases reversed..?

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: sry for confusion. status meant by active field in db

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: Technically if you refresh the page you shouldn't get different results because if the pod is up and running, the healthcheck should be performed automatically. Therefore, you should get fairly consistent result given that the clusters are not flaky.

However, there will still be inconsistent states if the clusters are flaky and the gateway is not picking up yet. 

I see the value of having a central place to store this info to avoid inconsistent states. While this is great, I think it might lead to some kind of race condition, and could potentially cause requests to be routed to an unhealthy cluster. 

For example, if gateway A sees cluster A is bad and tries to mark it unhealthy, in the meantime gateway B looks at DB and thinks cluster A is healthy, and hence routes requests to cluster A. 

If this information is only available locally, then each gateway will manage its own clusters' states. Then there won't be any race condition.

Any thoughts?

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: > Technically if you refresh the page you shouldn't get different results because if the pod is up and running, the healthcheck should be performed automatically.  Therefore, you should get fairly consistent results given that the clusters are not flaky.

True, this does not happen that often. 

> For example, if gateway A sees cluster A is bad and tries to mark it unhealthy, in the meantime gateway B looks at DB and thinks cluster A is healthy, and hence routes requests to cluster A.

If the DB has value, i don't see any cases where gateway A, B sees cluster A active status differently.

Or maybe have a lock field? 
Assuming cluster A is inactive, and cluster B active.
- Gateway A scans if there is any inactive cluster. Finds cluster A, check for lock, if there isn't any lock, locks cluster A
  - Update (querying to see number of nodes) and releases lock afteward.
- Gateway B, also scans, and checks for lock. Cluster A is locked so skip.

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: the current behavior is that if any cluster is **inactive** (meaning it's switched off manually), then the healthcheck will not be performed, since it doesn't make sense to perform healthcheck on **inactive** cluster.

If the cluster is **active**, then the healthcheck will perform. The cluster could be healthy, pending, or unhealthy. 

In your case, if Gateway A finds any inactive clusters, should skip them. It should only check on active clusters.

Comment (andythsu) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: The issue with storing healthcheck states in DB is that the healthcheck is done every minute and it's supposed to be a fast/small ping test. If we store this result in DB, we will need to look at DB everytime when healthcheck is performed, and hence it's adding more overhead. Also, when routingManager is choosing which cluster to route, it will need to look at DB to ensure the cluster is actually healthy. This could slow down the operation time.

**Comment by vishalya** [comment]
On a related topic, the pain point is that each gateway instance runs the background job for collecting the stats and the status from the trino clusters. IMO just one instance should do run the background job looking at the timestamp when the last one was run.

File: gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java (hunk 7 above)
Comment (willmostly) [reply gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/HealthCheckObserver.java:33]: I'll start a Slack thread to discuss design options

**Comment by andythsu** [comment]
Closing this PR for a new design: https://github.com/trinodb/trino-gateway/issues/222
